Content.Server/GameObjects/EntitySystems/Atmos/GasTileOverlaySystem.cs

```diff
@@ -0,0 +1,350 @@
+            
+            AccumulatedFrameTime -= UpdateTime;
+
+            var gridAtmosComponents = new Dictionary<GridId, GridAtmosphereComponent>();
+            var updatedTiles = new Dictionary<GasOverlayChunk, HashSet<MapIndices>>();
+           
+            // So up to this point we've been caching the updated tiles for multiple ticks.
+            // Now we'll go through and check whether the update actually matters for the overlay or not,
+            // and if not then we won't bother sending the data.
+            foreach (var (gridId, indices) in _invalidTiles)
+            {
+                var gridEntityId = _mapManager.GetGrid(gridId).GridEntityId;
+
+                if (!EntityManager.GetEntity(gridEntityId).TryGetComponent(out GridAtmosphereComponent gam))
+                {
+                    continue;
+                }
+
+                // If it's being invalidated it should have this right?
+                // At any rate we'll cache it for here + the AddChunk
+                if (!gridAtmosComponents.ContainsKey(gridId))
+                {
+                    gridAtmosComponents[gridId] = gam;
+                }
+
+                foreach (var invalid in indices)
+                {
+                    var chunk = GetOrCreateChunk(gridId, invalid);
+
+                    if (!TryRefreshTile(gam, chunk.GetData(invalid), invalid, out var data)) continue;
+
+                    if (!updatedTiles.TryGetValue(chunk, out var tiles))
+                    {
+                        tiles = new HashSet<MapIndices>();
+                        updatedTiles[chunk] = tiles;
+                    }
+                    
+                    updatedTiles[chunk].Add(invalid);
+                    chunk.Update(data, invalid);
+                }
+            }
+
+            var currentTime = _gameTiming.CurTime;
+
+            // Set the LastUpdate for chunks.
+            foreach (var (chunk, _) in updatedTiles)
+            {
+                chunk.Dirty(currentTime);
+            }
+
+            // Now we'll go through each player, then through each chunk in range of that player checking if the player is still in range
+            // If they are, check if they need the new data to send (i.e. if there's an overlay for the gas).
+            // Afterwards we reset all the chunk data for the next time we tick.
+            foreach (var (session, overlay) in _knownPlayerChunks)
+            {
+                if (session.AttachedEntity == null) continue;
+                
+                // Get chunks in range and update if we've moved around or the chunks have new overlay data
+                var chunksInRange = GetChunksInRange(session.AttachedEntity).ToArray();
+                var knownChunks = overlay.GetKnownChunks().ToArray();
+                var chunksToRemove = new List<GasOverlayChunk>(0);
+                var chunksToAdd = new List<GasOverlayChunk>(0);
```

Comment:
Tidbit: specifying 0 as capacity to a list does the exact same thing as not specifying capacity at all, [except it's slightly slower because it has to compare that you aren't passing a negative value etc.](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSAsA3OunAMy4LYDC2A3utk7hQDICWEwAPOwHbAAfNgCyACgCU9Rs1lwA7Nj4BTAO7YOXXgMFjME4mlkBfGUzMsNnHvyGiEk6UdnMFStVa23dBi6bTGQA)
```suggestion
                var chunksToRemove = new List<GasOverlayChunk>();
                var chunksToAdd = new List<GasOverlayChunk>();
```

Reply:
Cheers, good to know. I did this with the pathfinding system which I'll remember when I go back to it.